### PR TITLE
msg/async: remove useless judgement

### PR DESCRIPTION
**Pull request by majianpeng**

### CODE REVIEW

**Comment by majianpeng** [comment]
@gregsfortytwo . I'm not understand what's mean of  avoid doing operation.  I think those code on critical path and judgements are useless completely.

**Comment by gregsfortytwo** [comment]
I am embarrassed to say I don't remember any more if all messages here have contents in at least one of front, middle, data bufferlists. I looked a bit and I think at this point they do, but something used to only be in the header? Or maybe there is still one header-only message?

So if we have a message that doesn't need to read any more off the wire, invoking those throttlers (even for zero data) is more expensive than the if-check. We'll need to dig into that a little more, since I'm a wuss.

**Comment by majianpeng** [comment]
I see.  Thanks your detail explain. I'll drop a417eb4.

**Comment by gregsfortytwo** [comment]
Well this bit definitely looks fine. :)

File: src/msg/async/AsyncConnection.cc [hunk 1]
```diff
@@ -550,8 +550,7 @@ void AsyncConnection::process()
           // read front
           unsigned front_len = current_header.front_len;
           if (front_len) {
-            if (!front.length())
-              front.push_back(buffer::create(front_len));
+	    front.push_back(buffer::create(front_len));
```
Comment (runsisi) [inline src/msg/async/AsyncConnection.cc]: need one more space ?

File: src/msg/async/AsyncConnection.cc [hunk 2]
```diff
@@ -571,8 +570,7 @@ void AsyncConnection::process()
           // read middle
           unsigned middle_len = current_header.middle_len;
           if (middle_len) {
-            if (!middle.length())
-              middle.push_back(buffer::create(middle_len));
+	    middle.push_back(buffer::create(middle_len));
```
Comment (runsisi) [inline src/msg/async/AsyncConnection.cc]: ditto

File: src/msg/async/AsyncConnection.cc [hunk 3]
```diff
@@ -547,45 +547,41 @@ void AsyncConnection::process()
...
-            }
+	  // read front
+	  unsigned front_len = current_header.front_len;
+	  if (front_len) {
+	    front.push_back(buffer::create(front_len));
+	    r = read_until(front_len, front.c_str());
+	    if (r < 0) {
+	      ldout(async_msgr->cct, 1) << __func__ << " read message front failed" << dendl;
+	      goto fail;
+	    } else if (r > 0) {
```
Comment (tchaikov) [inline src/msg/async/AsyncConnection.cc]: could you avoid the s/<spaces>/<tab>/ in this commit?

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
I think this is part fo the logic for the state.. we only allocate teh buffer once, but may pass through this code multiple times.  This change seems to blow up in test.

**Comment by gregsfortytwo** [comment]
Oof, yeah, looking at this again, the change is forcibly appending a new buffer pointer of size front_len to the buffer. Previously we checked if a buffer already existed and only append one if it's not there yet. That doesn't make sense; I think we need all this code so I'm going to close the PR.

**Pull request closed without merging**